control-plane/agents/ha-node/src/path_provider.rs

```diff
@@ -0,0 +1,230 @@
+use crossbeam_queue::SegQueue;
+use futures_util::{future::ready, stream::StreamExt};
+use nvmeadm::nvmf_subsystem::{NvmeSubsystems, Subsystem};
+use std::{
+    collections::HashMap,
+    convert::TryInto,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+use tokio_udev::{AsyncMonitorSocket, EventType, MonitorBuilder};
+use utils::NVME_TARGET_NQN_PREFIX;
+
+const SYSFS_PREFIX: &str = "/sys/devices/virtual/nvme-fabrics/ctl/";
+/// Maximum number of events to process upon cache update (to prevent event flooding).
+const CACHE_EVENTS_BATCH_SIZE: usize = 128;
+
+#[derive(Debug, Clone)]
+pub struct NvmePath {
+    path_buffer: PathBuf,
+}
+
+/// Object that represents a path to sysfs NVMe object.
+impl NvmePath {
+    fn new(path_buffer: PathBuf) -> Self {
+        Self { path_buffer }
+    }
+
+    #[inline]
+    pub fn path(&self) -> &Path {
+        self.path_buffer.as_path()
+    }
+}
+
+/// Check that NVMe path represents a target created by our product and
+/// obtain a Path object that represents a system path for this NVMe device.
+fn get_nvme_path_buf(path: &String) -> Option<PathBuf> {
+    Path::new(path).canonicalize().ok().and_then(|pb| {
+        Subsystem::new(pb.as_path()).ok().and_then(|s| {
+            // Check NQN of the subsystem to make sure it belongs to the product.
+            if s.nqn.starts_with(NVME_TARGET_NQN_PREFIX) {
+                Some(pb)
+            } else {
+                None
+            }
+        })
+    })
+}
+
+/// UDEV-driven cache synchronization messages.
+#[derive(Debug, Clone)]
+enum CacheOp {
+    Add(String),
+    Remove(String),
+    InvalidateCache,
+}
+
+/// UDEV-based NVMe path provider that uses cache to keep all known NVMe targets
+/// in memory and avoid wildcard Sysfs enumeration when there is a need to get
+/// a list of all existing NVMe path in the system.
+/// This name provider sonsists of 2 parts: backend and frontend.
+/// Backend: (CachedNvmePathProvider itself) is responsible for monitoring UDEV events
+/// and signalling cache modification in case there are suitable UDEV events related
+/// to NVMe subsystem.
+/// Frontend: (NvmePathNameCollection), which represents a cache for all known NVMe
+/// paths in the system. Note: we only cache paths that represent NVMe targets with product's
+/// In order to eliminate cache inconsistency, there must be only one front-end path collection.
+/// NQN, so cache acts as a filter for 'unknown' NQNs.
+/// To eliminate locking/contention, backend communicates with frontend via a lockless queue.
+#[derive(Debug)]
+pub struct CachedNvmePathProvider {
+    udev_queue: Arc<SegQueue<CacheOp>>,
```

Comment:
Shouldn't SegQueue be protected by Mutex?

Reply:
It's a luckless queue which doesn't require locks.